File: public/docs/_examples/toh-6/ts/app/hero-search.service.ts
// #docregion
import { Injectable }     from '@angular/core';
import { Http, Response } from '@angular/http';

import { Hero }           from './hero';

@Injectable()
export class HeroSearchService {

  constructor(private http: Http) {}

  // #docregion observable-search
  search(term: string) {
    return this.http
               .get(`app/heroes/?name=${term}+`)
               .map((r: Response) => r.json().data as Hero[]);
  }
  // #enddocregion observable-search
}

File: public/docs/_examples/toh-6/dart/lib/hero_search_service.dart
// #docregion
import 'dart:async';
import 'dart:convert';

import 'package:angular2/core.dart';
import 'package:http/http.dart';

import 'hero.dart';

@Injectable()
class HeroSearchService {
  final Client _http;

  HeroSearchService(this._http);

  Future<List<Hero>> search(String term) async {
    try {
      final response = await _http.get('app/heroes/?name=$term');
      return _extractData(response)
          .map((json) => new Hero.fromJson(json))
          .toList();
    } catch (e) {
      throw _handleError(e);
    }
  }

  dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];

  Exception _handleError(dynamic e) {
    print(e); // for demo purposes only
    return new Exception('Server error; cause: $e');
  }
}
